File: web/docs-examples/typescript/examples/agent-tool-calling.ts
/**
 * Agent with tool calling — define tools, load agent prompty, invoke.
 *
 * Uses `tool()` to wrap functions with metadata and `bindTools()` to
 * validate handlers against the agent's declared tools.
 *
 * @example
 * ```bash
 * OPENAI_API_KEY=sk-... npx tsx examples/agent-tool-calling.ts
 * ```
 */
import "@prompty/openai";
import { tool, bindTools, turn, load } from "@prompty/core";
import { resolve } from "node:path";

const promptyFile = resolve(import.meta.dirname, "../../prompts/chat-agent.prompty");

// Define a tool with typed parameters
const getWeather = tool(
  (args: Record<string, unknown>) => {
    const city = args.city as string;
    return JSON.stringify({ city, temperature: 72, conditions: "sunny" });
  },
  {
    name: "get_weather",
    description: "Get the current weather for a city",
    parameters: [
      { name: "city", kind: "string", description: "City name", required: true },
    ],
  },
);

export async function agentToolCalling(question?: string): Promise<string> {
  const agent = load(promptyFile);

  // Validate tool handlers against agent's declared tools
  const tools = bindTools(agent, [getWeather]);

  // turn runs the LLM loop: call → tool dispatch → call → ...
  const result = await turn(agent, {
    question: question ?? "What's the weather in Seattle?",
  }, {
    tools,
    maxIterations: 5,
  });

  return result as string;
}

// Run directly
const response = await agentToolCalling();
console.log(response);

File: web/docs-examples/python/examples/agent_tool_calling.py
"""Agent with tool calling — register tools and run the agent loop.

This example shows how to define tools and run an agent that calls them.
Used in: how-to/agent-tool-calling.mdx
"""
from __future__ import annotations

from prompty import turn, load
from prompty.core import tool


@tool
def get_weather(city: str) -> str:
    """Get the current weather for a city."""
    return f"72°F and sunny in {city}"


agent = load("chat-agent.prompty")
result = turn(
    agent,
    inputs={"question": "What's the weather in Seattle?"},
    tools={"get_weather": get_weather},
)
print(result)
